src/CodeGeneration.Roslyn.Engine/DocumentTransform.cs

```diff
@@ -42,13 +42,15 @@ public static class DocumentTransform
         /// <param name="projectDirectory">The path of the <c>.csproj</c> project file.</param>
         /// <param name="assemblyLoader">A function that can load an assembly with the given name.</param>
         /// <param name="progress">Reports warnings and errors in code generation.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
         /// <returns>A task whose result is the generated document.</returns>
         public static async Task<SyntaxTree> TransformAsync(
             CSharpCompilation compilation,
             SyntaxTree inputDocument,
             string projectDirectory,
             Func<AssemblyName, Assembly> assemblyLoader,
-            IProgress<Diagnostic> progress)
+            IProgress<Diagnostic> progress,
+            CancellationToken cancellationToken = default)
```

Comment:
Is there any good reason the caller shouldn't provide one of these? So long as we're making breaking API changes as this does, should we mandate that they provide one?

Reply:
Well, to be frank I don't consider the `Engine` part to be a "public" API. But it is, obviously. So maybe we could just make it not be public?

On the actual issue: I think both `CancellationToken` and `IProgress` should still be pay-for-play. In most cases I've seen, they are optional. Or rather, I've not seen a case where they're required. But it's not a strong opinion, since as I've mentioned above, I'd not expect one to use the `Engine` package alone, and not consider it a supported scenario.

Reply:
I personally prefer that the cancellationtoken is mandatory. For me, cancellation is a crucial aspect of asynchronous behavior.
I am OK leaving at it is (PR), and/or making the `progress` optional, too.

Reply:
Well, let's try a different approach. Would you agree that required arguments are those without which the method cannot successfully execute?

And would you agree then that others could be called optional?

And finally, that the methods we discuss can execute *without* either CTS or Progress? :)

Reply:
Supporting `null` for `IProgress<T>` and `default(CancellationToken)` are different questions than whether we make the parameters themselves optional by providing default values.

I fully support any method working with null/None values for both of these parameter types.

However making a parameter *optional* makes it very easy for callers to neglect to provide arguments for them even when they are available. Making `cancellationToken` optional for example means that our (one?) caller might forget to pass one in, even though it might very well have a `cancellationToken` that _could_ be passed in!

So in general when designing APIs, I only make parameters optional if it's reasonable to assume that many callers will *legitimately* not want to pass in an argument. But if most/all of my callers should have a value for a parameter, I make it mandatory so they don't forget to provide it.

Reply:
I am lost in the discussion ;) How should the parameters be?

A) progress = null, cancellationToken = default
B) progress,  cancellationToken = default
C) progress,  cancellationToken

I prefer C :)

Reply:
@AArnott does "potential callers" include (unit) tests? Because in tests, it's very much legitimate to not want to provide those, and specifying those as "default" doesn't convey anything, it's just additional noise - that's how I see it.

That aside, I do understand the reasoning you provided. I've also looked at [.NET Guide for TAP](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap#cancellation-optional) and indeed there's no implied requirement for those parameters to be optional in the sense of syntax declarations, so I'm also going to *agree on making both parameters non-optional* syntax-wise.

Reply:
When it comes to public APIs, I do not optimize for tests. I optimize for expected actual users. That said, if my tests don't represent my typical users then there's something likely wrong with how I'm testing.

But most of my unit tests (not necessarily in this repo) *do* have `CancellationToken` available in order to help them gracefully timeout if they run too long. 

At the end of the day, I'm not too passionate either way on this question, both because the users of this API are few and because I think if a few of those callers do it wrong and omit the cancellation token, I'm not going to lose any sleep over it.

But yes, I'm suggesting option C.